### ISSUE TITLE
OBS Studio not working

### ISSUE THREAD

**Issue by vdegenne** [WayfireWM/wayfire#1802]
*(Sorry I deleted the template, but I made sure to include all the information that it asked)*

Obs studio doesn't propose any screen output source.
<img src="https://github.com/WayfireWM/wayfire/assets/2827383/be6888c1-f866-41db-a882-e9e57d5d5c84" width=325>

The reason I ask that here and not in obs-studio repository is because obs works in sway, I switched from sway to wayfire recently and this is the only inconvenience I noticed. I believe Obs uses something called pipewire in wayland environments.

**versions**
wayfire 0.7.4
obs 29.1.1 (64bit)

Here's the log when I run Obs from the command line, maybe it can help to debug:

```
debug: Found portal inhibitor
debug: Attempted path: share/obs/obs-studio/locale/en-US.ini
debug: Attempted path: /usr/share/obs/obs-studio/locale/en-US.ini
debug: Attempted path: share/obs/obs-studio/locale.ini
debug: Attempted path: /usr/share/obs/obs-studio/locale.ini
debug: Attempted path: share/obs/obs-studio/themes/Yami.qss
debug: Attempted path: /usr/share/obs/obs-studio/themes/Yami.qss
info: Platform: Wayland
info: CPU Name: AMD Ryzen 9 5900HS with Radeon Graphics
info: CPU Speed: 4466.433MHz
info: Physical Cores: 8, Logical Cores: 16
info: Physical Memory: 15381MB Total, 2761MB Free
info: Kernel Version: Linux 6.1.18-200.fc37.x86_64
info: Distribution: "Fedora Linux" 37
info: Desktop Environment: Wayfire (wayfire)
info: Session Type: wayland
info: Qt Version: 6.4.3 (runtime), 6.4.3 (compiled)
info: Portable mode: false
info: OBS 29.1.1 (linux)
info: ---------------------------------
info: ---------------------------------
info: audio settings reset:
        samples per sec: 48000
        speakers:        2
        max buffering:   960 milliseconds
        buffering type:  dynamically increasing
info: ---------------------------------
info: Initializing OpenGL...
info: Using EGL/Wayland
info: Initialized EGL 1.5
info: Loading up OpenGL on adapter AMD AMD Radeon Graphics (renoir, LLVM 15.0.7, DRM 3.49, 6.1.18-200.fc37.x86_64)
info: OpenGL loaded successfully, version 4.6 (Core Profile) Mesa 23.0.3, shading language 4.60
info: ---------------------------------
info: video settings reset:
        base resolution:   352x397
        output resolution: 352x396
        downscale filter:  Bicubic
        fps:               60/1
        format:            NV12
        YUV mode:          Rec. 709/Partial
info: NV12 texture support not available
info: P010 texture support not available
info: Audio monitoring device:
        name: Default
        id: default
info: ---------------------------------
warning: Failed to load 'en-US' text for module: 'decklink-captions.so'
warning: Failed to load 'en-US' text for module: 'decklink-output-ui.so'
libDeckLinkAPI.so: cannot open shared object file: No such file or directory
warning: A DeckLink iterator could not be created.  The DeckLink drivers may not be installed
warning: Failed to initialize module 'decklink.so'
info: [pipewire] No captures available
error: os_dlopen(libnvidia-encode.so.1->libnvidia-encode.so.1): libnvidia-encode.so.1: cannot open shared object file: No such file or directory

nvc0_screen_create:1169 - Error allocating PGRAPH context for M2MF: -110
error: VAAPI: Failed to initialize display in vaapi_device_h264_supported
info: VAAPI: API version 1.16
info: FFmpeg VAAPI H264 encoding not supported
nvc0_screen_create:1169 - Error allocating PGRAPH context for M2MF: -16
error: VAAPI: Failed to initialize display in vaapi_device_hevc_supported
info: FFmpeg VAAPI HEVC encoding not supported
info: [obs-websocket] [obs_module_load] you can haz websockets (Version: 5.2.2 | RPC Version: 1)
info: [obs-websocket] [obs_module_load] Qt version (compile-time): 6.4.3 | Qt version (run-time): 6.4.3
info: [obs-websocket] [obs_module_load] Linked ASIO Version: 102600
info: [obs-websocket] [obs_module_load] Module loaded.
info: [vlc-video]: VLC 3.0.19 Vetinari found, VLC video source enabled
info: ---------------------------------
info:   Loaded Modules:
info:     vlc-video.so
info:     text-freetype2.so
info:     rtmp-services.so
info:     obs-x264.so
info:     obs-websocket.so
info:     obs-vst.so
info:     obs-transitions.so
info:     obs-outputs.so
info:     obs-libfdk.so
info:     obs-filters.so
info:     obs-ffmpeg.so
info:     linux-v4l2.so
info:     linux-pulseaudio.so
info:     linux-pipewire.so
info:     linux-jack.so
info:     linux-capture.so
info:     linux-alsa.so
info:     image-source.so
info:     frontend-tools.so
info:     decklink-output-ui.so
info:     decklink-captions.so
info: ---------------------------------
QWidget::setTabOrder: 'first' and 'second' must be in the same window
info: ==== Startup complete ===============================================
info: All scene data cleared
info: ------------------------------------------------
info: [Python: resize-base-resolution-from-source2.py] <Swig Object of type 'obs_data_t *' at 0x7f03805b3ed0>
info: [obs-scripting]: Loaded python script: resize-base-resolution-from-source2.py
info: pulse-input: Server name: 'PulseAudio (on PipeWire 0.3.70) 15.0.0'
info: pulse-input: Audio format: s16le, 48000 Hz, 2 channels
info: pulse-input: Started recording from 'bluez_output.58_FC_C6_1C_55_4E.1.monitor' (default)
info: [Loaded global audio device]: 'Desktop Audio'
info: pulse-input: Server name: 'PulseAudio (on PipeWire 0.3.70) 15.0.0'
info: pulse-input: Audio format: s32le, 48000 Hz, 2 channels
info: pulse-input: Started recording from 'alsa_input.pci-0000_08_00.6.analog-stereo' (default)
info: [Loaded global audio device]: 'Mic/Aux'
error: Source ID 'pipewire-desktop-capture-source' not found
error: Failed to create source 'Screen Capture (PipeWire) 2'!
info: Switched to scene 'asdf'
info: ------------------------------------------------
info: Loaded scenes:
info: - scene 'asdf':
info: - scene 'sdf':
info:     - source: 'Screen Capture (PipeWire) 2' (pipewire-desktop-capture-source)
info: ------------------------------------------------
info: adding 21 milliseconds of audio buffering, total audio buffering is now 21 milliseconds (source: Desktop Audio)
```

**Comment by vdegenne** [comment]
@ammen99 I found this thread https://github.com/WayfireWM/wayfire/issues/1611 and decided to include 
```
[autostart]
01_xdp=/usr/libexec/xdg-desktop-portal
02_xdpg=/usr/libexec/xdg-desktop-portal-gtk
03_xdpw=/usr/libexec/xdg-desktop-portal-wlr
```
in my config file.

It seems that it works now! though I really don't know why. Does it change the way things are getting rendered on the screen in general?

(If there is a section in the documentation that talks about this, could you please redirect me?)

So far wayfire is really boosting my production and I can't be thankful enough for that, if there is anyway we can make donations, please let me know.

**Comment by ammen99** [comment]
Glad to hear it works :)

For screen recording/sharing/etc. Wayfire and most wlroots-based compositors (Sway too) have a semi-private protocol, called wlr-screencopy. However, after we had this protocol, another way for this was standardized, which is also supported by GNOME and KDE (via xdg-desktop-portal). There is `xdg-desktop-portal-wlr` which basically translates from the wlroots' protocol to the xdp one, and OBS will not work unless this bridging component runs in the background (since OBS uses the standard xdp recording protocol).

**Comment by vdegenne** [comment]
@ammen99 Well... I think it is somehow related but do you happen to know why file picker doesn't show up? For instance when I click "upload a file" in browser, a window should open so I can choose a file on my local machine, but for some reason it doesn't show up.

I thought autorun the portals would fix this issue but it doesn't.

**Comment by ammen99** [comment]
Do you run xdg-desktop-portal-gtk? It should be autostarted if you have the dbus/systemd env vars exported by autostart, the wiki has a section on that iirc.

**Comment by vdegenne** [comment]
@ammen99 yes here's the `[autostart]` section in my config file 
```
[autostart]
autostart_wf_shell = true
00_dbus = dbus-launch
01_xdp=/usr/libexec/xdg-desktop-portal
02_xdpg=/usr/libexec/xdg-desktop-portal-gtk
03_xdpw=/usr/libexec/xdg-desktop-portal-wlr
portal=/usr/libexec/xdg-desktop-portal-gtk
```
Maybe I shouldn't use all the portals at once?

Also I have the problem in chrome right now, but file chooser seems to work in Firefox. Maybe I should open a new issue for that?

**Comment by ammen99** [comment]
> Also I have the problem in chrome right now, but file chooser seems to work in Firefox. Maybe I should open a new issue for that?

Here's my config (no need for a new issue, this isn't a Wayfire issue anyway, it is all about the dbus stuff):

```[autostart]
0_0 = systemctl --user import-environment
portal = sleep 1 && (XDG_SESSION_TYPE=wayland XDG_CURRENT_DESKTOP=sway /usr/libexec/xdg-desktop-portal --replace & /usr/libexec/xdg-desktop-portal-wlr)
```

You can either set `XDG_*` before starting Wayfire, or replace the already-running portals like I do here.

You do need the systemctl command, or at least `dbus-update-activation-environment WAYLAND_DISPLAY DISPLAY XAUTHORITY` for all dbus-activated stuff to work, that includes file chooser dialogs.

**Comment by bodqhrohro** [comment]
Having same issue, I launched `xdg-desktop-portal-wlr`, restarted `pipewire`, but I still get the `info: [pipewire] No captures available` message on `obs` launch, and there are no relevant outputs in `helvum`.

**Comment by kode54** [comment]
To have the output picker show up, you need to have `slurp` installed.

**Comment by bodqhrohro** [comment]
I have it.

**Comment by kode54** [comment]
Really strange that it's not showing the Screen Capture (Pipewire) with the XDG Desktop Portal loaded.

**Comment by vdegenne** [comment]
Coming back to this issue, trying to find "window capture" (not "screen capture") but I can't find it, I don't know if that's a limitation of obs on wayland or another dbus issue.
Can you guys tell me if you have it?

**Comment by marcusbritanicus** [comment]
If I understand correctly, it's a limitation of wayland. There is currently, no way to identify a window. This is likely to change in the future, with the introduction of ext-foreign-toplevel-list protocol which provides a means to identify views with a unique id. This isn't yet supported by wlroots. Once the support for this is added, then you'll be able to capture views.

**Comment by soreau** [comment]
I think this might be the case of wlroots and wayfire supporting [this protocol](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/124), which outlines the ability to capture single toplevels.
